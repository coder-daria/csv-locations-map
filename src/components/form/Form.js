import React, { useState, useCallback } from 'react';
import arrayMove from 'array-move';

import SortableList from './sortable-list/SortableList';

import {
  ActionButtons,
  Button,
  Container,
  Info,
  StyledForm,
  SuccessMessage,
} from './Form.styles';

const FIELDS_NAMES = ['state', 'city', 'zip', 'address', 'category'];

function Form({ setColumnsOrder, fileName, rowsNum, uploadFile }) {
  const [fields, setFields] = useState(FIELDS_NAMES);
  const [isSorting, toggleSorting] = useState(false);

  const onSubmit = useCallback(
    () => {
        setColumnsOrder(fields);
    }, [fields, setColumnsOrder]
  );

  const onSortEnd = useCallback(
    ({ oldIndex, newIndex }) => {
      const orderedItems = arrayMove(fields, oldIndex, newIndex);

      setFields(orderedItems);
      toggleSorting(false);
    }, [fields],
  );

  const onSortStart = useCallback(
    () => {
      toggleSorting(true);
    }, [],
  );

  const cancelForm = useCallback(
    () => {
      uploadFile();
    }, [uploadFile],
  );

  return (
    <Container>
      <StyledForm>
        <SuccessMessage>
          <span>{fileName}</span>
          <span>{rowsNum} rows uploaded</span>
        </SuccessMessage>
        <Info>
          <span>
            To display locations correctly, drag and drop below fields to their representative column headers.
          </span>
        </Info>
        <SortableList
          isSorting={isSorting}
          axis='x'
          items={fields}
          onSortStart={onSortStart}
          onSortEnd={onSortEnd}
        />
        <ActionButtons>
          <Button onClick={cancelForm}>Cancel</Button>
          <Button onClick={onSubmit}>Accept</Button>
        </ActionButtons>
      </StyledForm>
    </Container>
  );
}

export default Form;
