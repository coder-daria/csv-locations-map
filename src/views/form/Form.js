import React from 'react';
import { func, string, number } from 'prop-types';

import SortableList from './sortable-list/SortableList';
import useFormDetails from './utils/use-form-details';

import { FORM_INFORMATION, BUTTON } from './constants';

import {
  ActionButtons,
  Button,
  Container,
  FileInformation,
  FormInfo,
  RowPreview,
  RowPreviewContainer,
  StyledForm,
} from './Form.styles';

const { ACCEPT, CANCEL } = BUTTON;

const SORTABLE_AXIS = 'x';

function Form({ setColumnsOrder, file, uploadFile }) {
  const {
    cancelForm,
    fields,
    isSorting,
    onSortEnd,
    onSortStart,
    onSubmit,
  } = useFormDetails({ setColumnsOrder, uploadFile });

  return (
    <Container>
      <StyledForm>
        <FileInformation>
          <span>{`${file.name} - ${file.data.length} rows`}</span>
        </FileInformation>
        <FormInfo>{FORM_INFORMATION}</FormInfo>
        <SortableList
          axis={SORTABLE_AXIS}
          isSorting={isSorting}
          items={fields}
          onSortEnd={onSortEnd}
          onSortStart={onSortStart}
        />
        <div>
          <div>First row in uploaded file.</div>
          <RowPreviewContainer>
            {file.data[0].map((rowItem) => <RowPreview>{rowItem}</RowPreview>)}
          </RowPreviewContainer>
        </div>
        <ActionButtons>
          <Button onClick={cancelForm}>{CANCEL}</Button>
          <Button onClick={onSubmit}>{ACCEPT}</Button>
        </ActionButtons>
      </StyledForm>
    </Container>
  );
}

Form.propTypes = {
  fileName: string.isRequired,
  rowsNum: number.isRequired,
  setColumnsOrder: func.isRequired,
  uploadFile: func.isRequired,
};

export default Form;
