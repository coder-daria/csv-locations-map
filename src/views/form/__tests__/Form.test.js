import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { FILE } from '../../__mocks__/file.mock';

import Form from '../Form';

const setup = () => {
  const props = {
    file: FILE,
    setColumnsOrder: jest.fn(),
    uploadFile: jest.fn(),
  };

  return shallow(<Form {...props} />);
};

describe('<Form />', () => {
  it('should match the sanpshot', () => {
    const wrapper = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
