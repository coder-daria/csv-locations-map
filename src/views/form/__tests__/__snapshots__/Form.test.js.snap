// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should match the sanpshot 1`] = `
<Styled(styled.div)>
  <styled.div>
    <styled.div>
      <span>
        locations.csv - 3 rows
      </span>
    </styled.div>
    <styled.span>
      To display locations correctly, drag and drop below fields to their representative column headers.
    </styled.span>
    <Memo(sortableList)
      axis="x"
      isSorting={false}
      items={
        Array [
          "state",
          "city",
          "zip",
          "address",
          "category",
        ]
      }
      onSortEnd={[Function]}
      onSortStart={[Function]}
    />
    <div>
      <div>
        First row in uploaded file.
      </div>
      <styled.div>
        <styled.span>
          Poland
        </styled.span>
        <styled.span>
          Wroclaw
        </styled.span>
        <styled.span>
          Igielna 19
        </styled.span>
        <styled.span>
          321
        </styled.span>
        <styled.span>
          Opera
        </styled.span>
      </styled.div>
    </div>
    <styled.div>
      <styled.button
        onClick={[Function]}
      >
        Cancel
      </styled.button>
      <styled.button
        onClick={[Function]}
      >
        Accept
      </styled.button>
    </styled.div>
  </styled.div>
</Styled(styled.div)>
`;
